sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1534,6 +1542,34 @@ object SimplifyCaseConversionExpressions extends Rule[LogicalPlan] {
 }
 
 /**
+ * Check if ther eany cartesian products between joins of any type in the optimized plan tree.
+ * Throw an error if a cartesian product is found without an explicit cross join specified.
+ * This rule is effectively disabled if the CROSS_JOINS_ENABLED flag is true.
```

Comment:
As discussed offline. Please document that we need to run this after join reordering.